smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java

```diff
@@ -0,0 +1,897 @@
+            remaining.remove(parseMember(id, remaining));
+            while (!eof()) {
+                ws();
+                if (charPeek() == ',') {
+                    expect(',');
+                    // A comma clears out any previously captured documentation
+                    // comments that may have been found when parsing the member.
+                    clearPendingDocs();
+                    ws();
+                    if (charPeek() == '}') {
+                        // Trailing comma: "," "}"
+                        break;
+                    }
+
+                    // Special casing to detect invalid members early on, even
+                    // after draining all the valid members. This keeps builders
+                    // from raising confusing errors about invalid members.
+                    if (remaining.isEmpty() && !requiredMembers.isEmpty()) {
+                        parseMember(id, requiredMembers);
+                    } else {
+                        remaining.remove(parseMember(id, remaining));
+                    }
+                } else {
+                    // Assume '}'; break to enforce.
+                    break;
+                }
+            }
+        }
+
+        if (!remaining.isEmpty()) {
+            throw syntax("Missing required members of shape `" + id + "`: ["
+                         + ValidationUtils.tickedList(remaining) + ']');
+        }
+
+        expect('}');
+    }
+
+    private String parseMember(ShapeId parent, Set<String> requiredMembers) {
+        // Parse optional member traits.
+        List<TraitEntry> memberTraits = parseDocsAndTraits();
+        SourceLocation memberLocation = currentLocation();
+        String memberName = IdlShapeIdParser.parseIdentifier(this);
+
+        // Only enforce "allowedMembers" if it isn't empty.
+        if (!requiredMembers.isEmpty() && !requiredMembers.contains(memberName)) {
+            throw syntax("Unexpected member of " + parent + ": '" + memberName + '\'');
+        }
+
+        ws();
+        expect(':');
+        ws();
+        ShapeId memberId = parent.withMember(memberName);
+        MemberShape.Builder memberBuilder = MemberShape.builder().id(memberId).source(memberLocation);
+        SourceLocation targetLocation = currentLocation();
+        String target = IdlShapeIdParser.parseShapeId(this);
+        visitor.onShape(memberBuilder);
+        onShapeTarget(target, targetLocation, memberBuilder::target);
+        addTraits(memberId, memberTraits);
+
+        return memberName;
+    }
+
+    private void parseMapStatement(ShapeId id, SourceLocation location) {
+        parseMembers(id, SetUtils.of("key", "value"));
+        visitor.onShape(MapShape.builder().id(id).source(location));
+    }
+
+    private void parseStructuredShape(ShapeId id, SourceLocation location, AbstractShapeBuilder builder) {
+        visitor.onShape(builder.id(id).source(location));
```

Comment:
There's inconsistency here in ordering of calls to `parseMembers` vs `visitor.onShape()`, is there significance that's not documented or is it just happenstance?

Reply:
Parsing members of list/set/map before registering the shape with the LoaderVisitor ensures that the shape is only registered if it as all of its required members. Otherwise, the LoaderVisitor gives a cryptic message with no context about how a "member" wasn't set on a builder. I added a comment to say this in code.

As for structures:

```
// Register the structure/union with the loader before parsing members.
// This will detect shape conflicts with other types (like an operation)
// and still give useful error messages. Trying to parse members first
// would otherwise result in cryptic error messages like:
// "Member `foo.baz#Foo$Baz` cannot be added to software.amazon.smithy.model.shapes.OperationShape$Builder"
```